zaino-common-foundation

nachog00 opened the pull request:
# Configuration Foundation: Shared Types and Parameter Cleanup

## Motivation

This PR addresses configuration architecture technical debt and establishes a foundation for shared configuration types across Zaino crates. The previous approach had scattered configuration parameters (`map_capacity`, `map_shard_amount`) that were performance tuning parameters exposed in consumer-facing config, making them difficult to maintain and error-prone (shard amounts had to be power-of-2).

## Solution

- **New `zaino-common` crate**: Centralized shared configuration types (`Network`, `ServiceConfig`, `StorageConfig`, `CacheConfig`, `DatabaseConfig`)
- **Modernized cache parameters**: Replaced raw `map_capacity`/`map_shard_amount` with structured `CacheConfig` containing `capacity` and `shard_power` (power-of-2 automatically enforced)
- **Network abstraction**: Unified network handling with conversion utilities between Zaino and Zebra network types
- **Type safety improvements**: `shard_power` as `u8` ensures valid power-of-2 values via `1 << shard_power`
- **Database defaults**: Set sensible 128GB default database size, removed unlimited option

### Tests

- All existing tests pass with new configuration structure
- Maintains backward compatibility for essential functionality 
- Database operations continue to work with new `StorageConfig`
- Network conversion utilities tested across Mainnet/Testnet/Regtest
- Config parsing tests updated with new structured fields (pending)

### Specifications & References

- DashMap performance tuning via capacity and shard configuration
- Zebra network types integration for seamless chain parameter access

### Follow-up Work

- Continue replacement of config fields into structs towards structure outlined in #467

### PR Checklist

- [X] The PR name is suitable for the release notes.
- [ ] The solution is tested.
- [X] The documentation is up to date.

nachog00 committed:
feat(zaino-commons): Add minimal shared configuration foundation

- Introduce zaino-commons crate for shared types across Zaino ecosystem
- Add Network enum with conversions to zebra-chain and zingo-infra types
- Add ServiceConfig for timeout and channel size settings
- Add storage configs: DatabaseConfig with explicit DatabaseSize enum, CacheConfig, StorageConfig
- Update workspace to include zaino-commons as first member

This foundation provides essential shared types that eliminate duplication
across zaino-state, zaino-serve, and other crates while maintaining
clean module organization for future growth.

nachog00 committed:
feat(zaino-state): Use zaino-commons shared configuration types

- Add zaino-commons dependency to zaino-state
- Refactor StateServiceConfig to use ServiceConfig and StorageConfig
- Refactor FetchServiceConfig to use shared configuration types
- Update BlockCacheConfig to use StorageConfig for cache/database settings
- Replace direct field access with structured config objects:
  - service_timeout/service_channel_size → service.timeout/service.channel_size
  - map_capacity/map_shard_amount → storage.cache.capacity/shard_amount
  - db_path/db_size → storage.database.path/size (with explicit DatabaseSize enum)
  - zebra_chain::parameters::Network → zaino_commons::Network with conversions
- Update all backend implementations to use new field structure

This demonstrates immediate value of zaino-commons by:
- Eliminating duplicated config fields across StateServiceConfig/FetchServiceConfig
- Providing type-safe DatabaseSize enum (Unlimited vs Limited{gb})
- Centralizing service timeout/channel configurations
- Creating reusable storage configuration pattern
- Adding network abstraction with automatic conversions

nachog00 committed:
Remove DatabaseSize::Unlimited, default to 128 GB

- Remove Unlimited variant from DatabaseSize enum for explicit resource management
- Default DatabaseSize to Gb(128) instead of unlimited growth
- Add as_bytes() method for GB to byte conversion (binary GB: 1024^3)
- Make DatabaseSize Copy for efficiency
- Add #[serde(default)] to DatabaseConfig.size field for TOML flexibility

TOML usage:
- size = { Gb = 256 } for explicit sizing
- size field can be omitted to use 128 GB default

nachog00 committed:
Make CacheConfig non-optional with defaults

- Remove Option<CacheConfig> from StorageConfig
- Add Default implementation for CacheConfig (capacity: 10000, shard_power: 4)
- Add #[serde(default)] to make cache field optional in TOML
- Users can omit cache field to get sensible defaults
- Simplifies API by eliminating None handling

nachog00 committed:
Simplify zaino-state config constructors

- Update StateServiceConfig::new() to take ServiceConfig and StorageConfig directly
- Update FetchServiceConfig::new() to take ServiceConfig and StorageConfig directly
- Update BlockCacheConfig::new() to take StorageConfig directly
- Remove individual parameter passing (map_capacity, map_shard_amount, db_path, db_size)
- Remove unused PathBuf import
- Eliminates parameter sprawl and uses structured config types

nachog00 committed:
Update database field access patterns

- Replace config.db_size with config.storage.database.size.as_bytes()
- Replace config.db_path with config.storage.database.path
- Use DatabaseSize.as_bytes() method for size conversion
- Eliminates direct field access in favor of structured config

nachog00 committed:
Fix network type conversions

- Use config.network.to_zebra_network() where zebra_chain::Network is expected
- Replace config.network.kind() with config.network.to_zebra_network().kind()
- Update ChainBlock::try_from calls to use converted network type
- Fixes type mismatches between zaino_commons::Network and zebra_chain::Network

nachog00 committed:
Update local cache field access patterns

- Replace config.db_size with config.storage.database.size.as_bytes()
- Replace config.db_path with config.storage.database.path
- Update LMDB map_size to use calculated bytes directly
- Add network conversions for fetch_block_from_node calls
- Consistent with structured config access patterns

nachog00 committed:
Export DatabaseSize type from zaino-commons

nachog00 committed:
Fix remaining compilation errors in foundation branch

- Fix all network type conversion issues by using .to_zebra_network()
- Update field access patterns for new config structure
- Fix cache initialization with proper Option wrapping
- Ensure all ChainBlock conversions use zebra network types

nachog00 committed:
rename zaino-commonS -> zaino-common

nachog00 committed:
Rename "zaino-commons" to "zaino-common" across files
- Updated Cargo.lock: corrected package name from "zaino-commons" to "zaino-common"
- Updated Cargo.toml: member name corrected to "zaino-common"
- Updated zaino-common/Cargo.toml: package name updated
- Updated zaino-state/Cargo.toml: dependency path corrected to "zaino-common"
- Updated zaino-state/src/config.rs: corrected import statement

This change ensures consistency in package naming across the repository.

nachog00 committed:
Merge branch 'dev' into zaino-commons-foundation

nachog00 commented:
## Configuration Architecture Improvements - Relationship to Config Issues

This PR begins addressing several configuration antipatterns identified in our Config-labeled issues. While it doesn't fully resolve these systemic issues, it establishes a foundation and demonstrates the path forward.

### Issues Directly Addressed

** #508 - Power-of-Two Constraint - Cache shard count causes runtime panics**
- **Resolved**: Replaced raw `map_shard_amount: Option<usize>` with `shard_power: u8` in `CacheConfig`
- **Solution**: `1 << shard_power` automatically ensures power-of-2 values, eliminating runtime panics
- **Impact**: Type-level safety prevents invalid shard configurations

**=� #500 - Primitive Obsession - Configuration uses primitive types requiring runtime validation**
- **Partially Addressed**: Introduced structured types (`StorageConfig`, `CacheConfig`, `ServiceConfig`)
- **Progress**: Replaced primitive cache parameters with proper structs containing semantic meaning
- **Remaining**: `IndexerConfig` still uses primitive types (string network, raw booleans, etc.)

**=� #499 - Configuration Shotgun Surgery - Duplicated fields across config structs**
- **Partially Addressed**: Created shared `zaino-common` crate with reusable config types
- **Progress**: `StorageConfig` and `ServiceConfig` now shared between `StateServiceConfig` and `FetchServiceConfig`
- **Remaining**: `IndexerConfig` still duplicates patterns, hasn't been updated to use shared types

### Issues Partially Addressed

**=� #515 - Inappropriate Intimacy - Public API tightly coupled to Zebra's internal types**
- **Progress**: Created `zaino_common::Network` with conversion utilities to Zebra types
- **Progress**: New configs use Zaino's `Network` type instead of Zebra's directly
- **Remaining**: Some lower-level library configs still expose `zebra_chain::parameters::Network` publicly
- **Remaining**: `IndexerConfig` conversion still directly creates Zebra network types

**=� #507 - Stringly-Typed Configuration - Using strings where enums would be safer**
- **Progress**: Created proper `zaino_common::Network` enum to replace network strings
- **Critical Gap**: `IndexerConfig` still uses `pub network: String` - not updated to use new enum!
- **Impact**: Main consumer-facing config still suffers from string typing issues

**=� #502 - God Object - IndexerConfig mixes concerns across multiple domains**
- **Progress**: Extracted storage and service concerns into structured subtypes for internal configs
- **Critical Gap**: `IndexerConfig` itself unchanged - still mixes TLS, auth, network, cache, database concerns
- **Remaining**: Consumer-facing config structure needs the same treatment

**=� #505 - Implicit Invariants - Cache configuration relies on undocumented constraints** 
- **Progress**: Made power-of-2 constraint explicit through `shard_power` type design in internal configs
- **Remaining**: `IndexerConfig` still has `map_capacity`/`map_shard_amount` with implicit constraints

**=� #501 - Telescoping Constructor - Config constructors with 15+ parameters**
- **Progress**: New structured config types reduce parameter count in internal construction
- **Remaining**: `IndexerConfig` constructors and conversion logic still telescoping

### Issues Not Yet Addressed (Future Work)

**L #506 - Long Method - Config validation contains complex branching logic**
- **Status**: `IndexerConfig::check_config()` validation logic remains complex and monolithic
- **Future**: Domain-specific validation methods per concern area

**L #504 - Feature Envy - Config structs manipulate external state instead of owning behavior**
- **Status**: Configs still primarily data containers
- **Future**: Move behavior into config types (validation, conversion, etc.)

**L #503 - Boolean Blindness - Authentication uses bool flags allowing invalid states**
- **Status**: Still using `validator_cookie_auth: bool`, `grpc_tls: bool` etc. in `IndexerConfig`
- **Future**: Enum-based authentication and TLS configuration

### Critical Missing Piece: IndexerConfig Updates

**The main consumer-facing `IndexerConfig` hasn't been updated to use any of the new structured types!** This is a significant oversight that should be addressed:

1. **Network**: Should use `zaino_common::Network` instead of `String`
2. **Cache/Storage**: Should use structured `StorageConfig` instead of raw primitives
3. **Services**: Should use `ServiceConfig` for timeout/channel settings
4. **Structure**: Should be decomposed into domain-specific sub-configs

### Foundation for Future Improvements

This PR establishes `zaino-common` as the foundation, but **we need to actually update `IndexerConfig` to use these new types** to realize the benefits. The internal config improvements show the path forward, but the main consumer API needs the same treatment.

**Immediate Next Steps**:
1. **Update `IndexerConfig`** to use `zaino_common::Network` instead of `String` 
2. **Integrate `StorageConfig`** into `IndexerConfig` to replace primitive cache fields
3. **Continue domain separation** per #467 for TLS, auth, service configs

idky137 commented on zaino-common/src/network.rs:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    /// Get the standard regtest activation heights used by Zaino.
+    pub fn zaino_regtest_heights() -> zebra_chain::parameters::testnet::ConfiguredActivationHeights
```
Would it be possible to add a separate method for setting custom regtest heights? This is sometimes useful for testing so if possible to add here could  be good.

AloeareV replied:
> Would it be possible to add a separate method for setting custom regtest heights? (not ) This is sometimes useful for testing so if possible to add here could be good.

Not what? I think you started to type a parenthetical here and didn't finish

nachog00 replied:
we thought of adding the heights as fields on the Testnet variant, so they can be passed from the config. It will surely not be part of this PR to keep it simpler. But it wouldn't be too hard to add later on. Good think is all that logic we can keep tight and together on the Network enum and aleviate the cosnumer components from dealing with it.

idky137 replied:
I think having them on Regtest is more useful than Testnet. Currently the only custom Testnet we use is Regtest, with Testnet internally meaning the public "Testnet".

This could be useful in the future but having a configurable Regtest would be more useful in the short term.

idky137 replied:
Ah the (not) was meant to say: Not necessarily exposed in the actual config file, but accessible internally for test usage. eg TestManager..

nachog00 replied:
> I think having them on Regtest is more useful than Testnet. Currently the only custom Testnet we use is Regtest, with Testnet internally meaning the public "Testnet".
> 
> This could be useful in the future but having a configurable Regtest would be more useful in the short term.

my bad, I meant regtest variant

idky137 commented on zaino-common/src/storage.rs:
```diff
@@ -0,0 +1,102 @@
+
+impl DatabaseSize {
+    /// Convert to bytes
+    pub fn as_bytes(&self) -> usize {
```
Could we update this name to  `as_byte_count` or something similar? 

`as_bytes` methods are generally used to convert to a sequence of raw u8 bytes so I think this could be confusing.

idky137 commented on zaino-common/src/network.rs:
```diff
@@ -0,0 +1,112 @@
+    ///
+    /// - Mainnet/Testnet: Skip sync (false) because we don't want to sync real chains in tests
+    /// - Regtest: Enable sync (true) because regtest is local and fast to sync
+    pub fn should_sync_for_testing(&self) -> bool {
```
Is this to replace the `no_sync` setting used for regtest connection?

If so then I think there is a slight misunderstanding of what the original bool did (probably because the docs were not clear).

What that bool actually did was told the `BlockCache` not to wait for the server to sync.

When running on mainnet or testnet the old `BlockCache` code used the `Blocks` (the amount of blocks the local validator has synced) and the `estimated height` (the estimated height of the zcash network, calculated by the local validator) fields from `get_blockchain_info` to determine when the validator was synced, and therefore when to start service.

With regtest this did not work as the estimated height was always incorrect, and due to the fact there is no network to sync with we used this bool to completely remove this check.

Currently the new `ChainIndex` code does not use this at all, it currently uses a different set of notifiers that should not be effected in the same way.

However we are still in the process of integrating this code so may still need to use something similar so could be good to get in. 

Given what this is doing (if i have read it correct) I think a more suitable name may be "wait_on_server_sync" or something similar.

nachog00 committed:
Refactor configuration to use zaino-common structs

- Cargo.lock and Cargo.toml
  - Added `zaino-common` as a dependency
- config.rs
  - Replaced primitive network string with `Network` enum
  - Migrated to `ServiceConfig` and `StorageConfig` from `zaino-common`
  - Removed manual validation of network type, relying on `Network` enum
  - Simplified network retrieval in `get_network()` using `Network` enum method
  - Replaced manual field declaration with default methods in `IndexerConfig::default()`
  - Updated `BackendConfig` conversion to use new configurations

This refactor uses structured data models to improve clarity and reduce manual string and value management.

AloeareV committed:
Merge branch 'dev' into zaino-commons-foundation

AloeareV committed:
still noncompiling work on commons

AloeareV committed:
get it all compiling

AloeareV committed:
rename

AloeareV committed:
remove some provided defaults in config test and fix database path and comment

AloeareV committed:
fix whitespace

AloeareV committed:
clippify

AloeareV committed:
add back used new function

AloeareV committed:
format

AloeareV committed:
fix hyperlink

idky137 commented on zaino-common/src/storage.rs:
```diff
@@ -0,0 +1,99 @@
+///
+/// This is used by services that need both in-memory caching and persistent storage.
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize, Default)]
+pub struct StorageConfig {
```
Non blocking, just thought I'd point out that `CacheConfig` here is only related to the in-memory cache rather than on disc storage so `StorageConfig` may not be completely accurate.

Im happy for this to land and for us to look at some of these names at a later date if best but, afaict it will mean an external, client facing config change, so we should confirm this name before release.

nachog00 replied:
agreed. We also had some chats about possibly moving some of these behind a 'dev' feature

idky137 commented on zaino-state/src/chain_index/tests/finalised_state/v0.rs:
```diff
@@ -22,26 +24,16 @@ pub(crate) async fn spawn_v0_zaino_db(
-        ),
+            ..Default::default()
+        },
+        db_version: 1,
```
```suggestion
        db_version: 0,
```

idky137 commented on zaino-state/src/chain_index/tests/finalised_state/v0.rs:
```diff
@@ -212,26 +204,16 @@ async fn load_db_from_file() {
-        ),
+            ..Default::default()
+        },
+        db_version: 1,
```
```suggestion
        db_version: 0,
```

idky137 reviewed (changes requested):
Looks good! 

Just the 2 small changes in the v0 ZainoDB tests. One was already present before this PR (slipped through original impl and review) but would be good to fix both of these and is only a one line change.

Also it looks like there are a couple failing config tests now, it may be that the actual tests need updating.

AloeareV committed:
fix remaining config tests

AloeareV committed:
fix v1 in v0 tests

idky137 committed:
Merge branch 'dev' into zaino-commons-foundation
